File: elixir/acronym/lib/acronym.ex
defmodule Acronym do
  @doc """
  Generate an acronym from a string.
  "This is a string" => "TIAS"
  """
  @spec abbreviate(String.t()) :: String.t()
  def abbreviate(string) do
    do_abbr(String.upcase(string) |> String.to_charlist(), :alpha, '')
  end

  # This is implemented with in little state machine.
  # Iterating over the string character-by-character:
  #
  # - in `:alpha` state, we skip over characters until we find a letter, then:
  #     - add the letter to the acronym
  #     - change the state to `:nonalpha`
  # - in `:nonalpha` state, we skip over characters that _are_ letters until
  #   we find a "non-alpha" (apostrophe are considered letters in this state)
  #     - change the state to `:alpha`

  @type state :: :alpha | :nonalpha

  @spec do_abbr(chars :: charlist, state :: state, abbr :: charlist) :: String.t()

  defp do_abbr([], _, abbr), do: to_string(abbr)

  defp do_abbr([char | chars], :alpha, abbr) when char in ?A..?Z do
    do_abbr(chars, :nonalpha, abbr ++ [char])
  end

  # apostrophes are to be ignored
  defp do_abbr([?' | chars], :nonalpha, abbr) do
    do_abbr(chars, :nonalpha, abbr)
  end

  defp do_abbr([char | chars], :nonalpha, abbr) when char not in ?A..?Z do
    do_abbr(chars, :alpha, abbr)
  end

  defp do_abbr([_ | chars], state, abbr), do: do_abbr(chars, state, abbr)
end

File: clojure/acronym/src/acronym.clj
(ns acronym
  (:require [clojure.string :as str]))

;; Two implementations:
(declare acronym-regex
         acronym-recursive)

(defn acronym
  "Converts phrase to its acronym."
  [phrase]
  ;(acronym-regex phrase)
  (acronym-recursive phrase)
)

;; ==============================================
;; using regular expressions
(defn title-case
  "Upper-case all word-starting letters"
  [phrase]
  (str/replace phrase #"\b\p{Lower}" #(str/upper-case %1)))

(defn acronym-regex [phrase]
  ;; find all upper-case letters that are _preceded_ by 
  ;; the start-of-string or a non-upper-case character.
  (->> (title-case phrase)
       (re-seq #"(?<=^|\P{Upper})\p{Upper}")
       str/join))

;; ==============================================
;; iterating over the characters
(defn upper? [chr] (<= 65 (int chr) 90))   ;; A to Z
(defn lower? [chr] (<= 97 (int chr) 122))  ;; a to z
(defn alpha? [chr] (or (upper? chr) (lower? chr)))

(defn acronym-letter? [chr prev]
  (or
    (and (alpha? chr) (not (alpha? prev)))
    (and (upper? chr) (lower? prev))))

(defn acronym-recursive 
  ([phrase]
    (acronym-recursive (seq phrase) \0 ""))
  ([chrs prev acronym]
    (if (empty? chrs)
      (str/upper-case acronym)
      (let [chr (first chrs)]
        (if (acronym-letter? chr prev)
          (recur (rest chrs) chr (str acronym chr))
          (recur (rest chrs) chr acronym))))))
